### DISCUSSION TITLE
What is the relationship and differences of two segments in DataNode & QueryNode

### DISCUSSION THREAD

**Discussion by Reidddddd** [milvus-io/milvus#37073] [Q&A and General discussion] [answered]
As far as I known, DataNode will handle writing data into object store by `writeNode` in a channel pipeline where there's a syncSegment call.

But I'm a bit confused when reading the source codes of QueryNode when I found there's a `insertNode` also will new segment and insert data into it. (seems to be local segment?)

Here we back to the question as QA title.

Hoping can get some insight answers from community. Thx.

**Comment by yhmo** [comment]
A "growing segment" exists both in query node and data node.
Query node subscribes data from message queue, the newly insert data is accumulated in the "growing segment".
Data node also subscribes data from message queue, the newly insert data also is accumulated inthe "growing segment".
The "growing segment" in query node and data node are pointing to the same logic segment.

Once the size of the growing segment exceeds a threshold value, the data node will "seal" it and it becomes a "sealed segment". At the same moment, the query node will know it become a sealed segment.

**Reply by Reidddddd** [reply]
` are pointing to the same logic segment.`

Is that mean, physically they are different?

**Reply by Reidddddd** [reply]
If physically different, then where is the segment inside Querynode stored?

**Reply by yhmo** [reply]
You create a collection, call collection.load() previously, and then insert 10 rows of data.
There is only one logic segment in the collection, the segment has 10 rows. The segment is "growing" since it only has 10 rows.
The query node has one copy of the segment in memory.
The data node also has one copy of the segment in memory.

You continue to insert 5 rows of data.
The query node receives the 5 rows from the message queue, now the querynode's growing segment has 15 rows.
The data node also receives the 5 rows from the message queue, now the datanode growing segment has 15 rows.

Query node doesn't persist the segment, it only keeps it in memory for search/query.
Data node doesn't do any search/query, it just keeps the segment in memory until its size exceeds the threshold. One exceed the threshold, the data node will persist the segment into storage as a "seal segment", and notify the query node "this segment is now converted to be sealed, you can sync its data to this timepoint xxxx". Then the query node knows "ok, sync the data to this timepoint xxxx and mark this segment as sealed".

**Comment by Reidddddd** [comment]
Screenshots are querynode's parts
![Screenshot 2024-10-23 at 15 30 06](https://github.com/user-attachments/assets/f9805cbc-ce92-46ef-bf2c-7c3d19fce14e)
![Screenshot 2024-10-23 at 15 30 47](https://github.com/user-attachments/assets/bfd73107-1d42-48b8-bf82-fcbac4bbdd8e)

**Reply by Reidddddd** [reply]
Can tell it create a segment on its own, which is not stored in object store

but hard for me to track since it goes into C++ parts.

 (:( Goland haven't supported golang/c++ mixed complier..

**Reply by Reidddddd** [reply]
> Screenshots are querynode's parts ![Screenshot 2024-10-23 at 15 30 06](https://private-user-images.githubusercontent.com/5352837/379150182-f9805cbc-ce92-46ef-bf2c-7c3d19fce14e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjk2NzUwODYsIm5iZiI6MTcyOTY3NDc4NiwicGF0aCI6Ii81MzUyODM3LzM3OTE1MDE4Mi1mOTgwNWNiYy1jZTkyLTQ2ZWYtYmYyYy03YzNkMTlmY2UxNGUucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTAyMyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDEwMjNUMDkxMzA2WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NWJkMGFkYWM5MjdkNzc0ZmJjYzI0MWM2ZGQ4ZDM2MDNlZWYyZjc5NjM4NzM2MDgzOTQyNWY5NmYzZDY0ZjA3ZSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.PUsRuRTDL2u5iqWHaRZK4nymom2esE_-USLApbiVips) ![Screenshot 2024-10-23 at 15 30 47](https://private-user-images.githubusercontent.com/5352837/379150202-bfd73107-1d42-48b8-bf82-fcbac4bbdd8e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjk2NzUwODYsIm5iZiI6MTcyOTY3NDc4NiwicGF0aCI6Ii81MzUyODM3LzM3OTE1MDIwMi1iZmQ3MzEwNy0xZDQyLTQ4YjgtYmY4Mi1mY2JhYzRiYmRkOGUucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTAyMyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDEwMjNUMDkxMzA2WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZWRmM2EyMjVhZDZkYmVlMzk3ZjVkZjlkMmM5ZWZhZDMzMWQ4MDJlOWRmZDhmNTM4NDYyYWM0MmEyZmQ0MzVmNSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.wE-MnSMknZHd1H10npzB1PoW1FKjJ-vV6gIwwWqDE88)

@yhmo 
from you sayings, these parts, in fact, is in memory.

**Reply by yhmo** [reply]
I believe this function is the query node receive data from the message queue and insert the data into the in-memory growing segment.
Yes, it is in memory, for query node.

```
func (sd *shardDelegator) ProcessInsert(insertRecords map[int64]*InsertData) {
      growing, err = segments.NewSegment()
      .......
      err := growing.Insert(context.Background(), insertData.RowIDs, insertData.Timestamps, insertData.InsertRecord)
}
```

**Reply by Reidddddd** [reply]
> I believe this function is the query node receive data from the message queue and insert the data into the in-memory growing segment. Yes, it is in memory, for query node.
> 
> ```
> func (sd *shardDelegator) ProcessInsert(insertRecords map[int64]*InsertData) {
>       growing, err = segments.NewSegment()
>       .......
>       err := growing.Insert(context.Background(), insertData.RowIDs, insertData.Timestamps, insertData.InsertRecord)
> }
> ```

yes it is, thx!

May I ask why this part of memory needs to be written in C++, any advantage concern?

**Reply by liliu-z** [reply, marked as answer]
> > I believe this function is the query node receive data from the message queue and insert the data into the in-memory growing segment. Yes, it is in memory, for query node.
> > ```
> > func (sd *shardDelegator) ProcessInsert(insertRecords map[int64]*InsertData) {
> >       growing, err = segments.NewSegment()
> >       .......
> >       err := growing.Insert(context.Background(), insertData.RowIDs, insertData.Timestamps, insertData.InsertRecord)
> > }
> > ```
> 
> yes it is, thx!
> 
> May I ask why this part of memory needs to be written in C++, any advantage concern?

In the design, C++ is the data plane and Go is the control plane. All index building and searching work of vector/scalar data are in C++ side to make it more efficient, so to avoid complicated data transfer between Go and C++, we decide to put all data processing work in the C++ side

**Comment by Reidddddd** [comment]
Thanks @yhmo @liliu-z 🙏
